dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/dataelement/DataElementController.java

```diff
@@ -27,21 +27,64 @@
  */
 package org.hisp.dhis.webapi.controller.dataelement;
 
-import lombok.AllArgsConstructor;
+import static org.hisp.dhis.security.Authorities.F_DATA_ELEMENT_MERGE;
+import static org.hisp.dhis.webapi.controller.CrudControllerAdvice.getHelpfulMessage;
+import static org.springframework.http.MediaType.APPLICATION_JSON_VALUE;
+
+import javax.persistence.PersistenceException;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
 import org.hisp.dhis.dataelement.DataElement;
 import org.hisp.dhis.dataelement.DataElementService;
+import org.hisp.dhis.dxf2.webmessage.WebMessage;
+import org.hisp.dhis.dxf2.webmessage.WebMessageUtils;
+import org.hisp.dhis.feedback.ConflictException;
+import org.hisp.dhis.feedback.MergeReport;
+import org.hisp.dhis.merge.MergeParams;
+import org.hisp.dhis.merge.MergeProcessor;
+import org.hisp.dhis.merge.MergeType;
+import org.hisp.dhis.security.RequiresAuthority;
 import org.hisp.dhis.webapi.controller.AbstractCrudController;
+import org.springframework.http.HttpStatus;
 import org.springframework.stereotype.Controller;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
 import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.ResponseBody;
+import org.springframework.web.bind.annotation.ResponseStatus;
 
 /**
  * @author Morten Olav Hansen <mortenoh@gmail.com>
  */
+@Slf4j
 @Controller
 @RequestMapping("/api/dataElements")
-@AllArgsConstructor
+@RequiredArgsConstructor
 public class DataElementController extends AbstractCrudController<DataElement> {
+
   private final DataElementService dataElementService;
+  private final MergeProcessor dataElementMergeProcessor;
+
+  @ResponseStatus(HttpStatus.OK)
+  @RequiresAuthority(anyOf = F_DATA_ELEMENT_MERGE)
+  @PostMapping(value = "/merge", produces = APPLICATION_JSON_VALUE)
+  public @ResponseBody WebMessage mergeDataElements(@RequestBody MergeParams params)
+      throws ConflictException {
+    log.info("Data element merge received");
```

Comment:
Left over debug line?

Reply:
nope. Purposely left in. The logging in our system could be much better. We should be logging more, especially when it's not something that will be spamming the logs. It's great to see what's going on in the logs. Nothing worse that doing something in the system, checking the logs, only to see nothing has changed.

Reply:
Generally I agree we could think more about where more logging makes sense. Not sure I find controller layer a good choice. Feels like that info should be possible to get with the right log level settings from the general HTTP container logs that I would assume are made by jetty or tomcat. but lets not make this complicated. we keep it and talk about the general log strategy in one of our meetings.